icu4c/source/i18n/calendar.cpp

```diff
@@ -733,10 +733,10 @@ fTime(0),
 fLenient(true),
 fZone(nullptr),
 fRepeatedWallTime(UCAL_WALLTIME_LAST),
-fSkippedWallTime(UCAL_WALLTIME_LAST)
+fSkippedWallTime(UCAL_WALLTIME_LAST),
+validLocale(new CharString()),
+actualLocale(new CharString())
 {
-    validLocale[0] = 0;
-    actualLocale[0] = 0;
     if (U_FAILURE(success)) {
         delete zone;
```

Comment:
optional: in order to track the adopted zone, we could stick it into a LocalPointer at the start of the constructor body, and orphan() it when we copy the pointer into fZone